crates/subspace-farmer/src/bin/subspace-farmer/utils.rs

```diff
@@ -1,4 +1,50 @@
+use std::future::Future;
 use std::path::PathBuf;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+use tokio::task::JoinHandle;
+
+type OnExitHandler = std::pin::Pin<Box<dyn Future<Output = ()> + Send>>;
+
+pub(crate) struct SignalHandler {
+    on_exit: Arc<Mutex<Vec<OnExitHandler>>>,
+    handle: Option<JoinHandle<std::io::Result<()>>>,
+}
+
+impl Drop for SignalHandler {
+    fn drop(&mut self) {
+        if let Some(handle) = self.handle.take() {
+            let _ = tokio::runtime::Handle::current().block_on(handle);
+        }
+    }
+}
+
+impl SignalHandler {
+    pub fn new() -> Self {
+        let on_exit = Arc::default();
+        Self {
+            on_exit: Arc::clone(&on_exit),
+            handle: Some(tokio::spawn(async move {
+                tokio::signal::ctrl_c().await?;
```

Comment:
Hm... I don't really understand why all this is necessary. Couldn't you just race this with what farmer normally does instead?